### PR TITLE
FloatMatrix with OwnedBuffer backend.

### PR DESCRIPTION
**Pull request by simonmika**
Added a OwnedBuffer backend to FloatMatrix so that it can use the notion of ownership to simplify freeing.

### CODE REVIEW

**Comment by simonmika** [comment]
@tomasmahlberg: we should try using this together for a start and see how it works in real usage.

**Comment by tomasmahlberg** [comment]
Alright. This looks nice.

**Comment by tomasmahlberg** [comment]
I've been experimenting a bit with this. Let's say we want to do the following (assuming `x` and `y` are two square `FloatMatrix`'s):
```ooc
for (i in 0 .. x width)
	x += 3.0f * y getColumn(i) * y getColumn(x width - 1 - i) transpose() + y
```
This will generate six memory leaks per iteration and would, if I understood this correctly, have to be rewritten as:
```ooc
for (i in 0 .. x width)
	x += ((3.0f * (y getColumn(i) give() * y getColumn(x width - 1 - i) give() transpose() give()) give() + y) give()) give()
```
This still looks a bit messy and should probably be split up onto more than one line. Would it make sense to instead let `give()` be the default behaviour in `FloatMatrix`'s methods, i.e. to have them return `result give()` instead of `result`? What consequences would this have?

**Comment by simonmika** [comment]
Thats what I want to look at.

**Comment by davidhesselbom** [comment]
> Would it make sense to instead let give() be the default behaviour in FloatMatrix's methods, i.e. to have them return result give() instead of result?

That sounds a lot like doing things in-place instead of making `FloatMatrix` immutable, which I what I suggested earlier...

**Comment by davidhesselbom** [comment]
Anyway, is this still ready to be merged, despite the messy syntax?

**Comment by simonmika** [comment]
An update to the syntax is coming. @tomasmahlberg can you do a PR on my branch?

**Comment by tomasmahlberg** [comment]
There's still the problem with free being called multiple times in operator methods when instantiating a `FloatMatrix` with owner `Owner Callee` instead of `Owner Caller` as I have not had time to work on this yet. Merge this PR as is, and fix this later?

**Comment by simonmika** [comment]
Better not to merge before fixing as it, most likely, will break existing code.

File: source/math/FloatMatrix.ooc [hunk 1]
```diff
@@ -219,44 +224,27 @@ FloatMatrix : cover {
...
-				temp1 free()
-				temp2 free()
-				lup[0] free()
-				lup[1] free()
-				lup[2] free()
-				lup free()
+				outerProduct := (this transpose() * this) take()
+				(l, u, p) := outerProduct lupDecomposition()
+				result = (p * this transpose() * y) forwardSubstitution(l) backwardSubstitution(u)
+				outerProduct free()
```
Comment (davidhesselbom) [inline source/math/FloatMatrix.ooc]: Wow, nice!

File: source/math/FloatMatrix.ooc [hunk 2]
```diff
@@ -24,25 +25,26 @@ FloatMatrix : cover {
 	dimensions ::= this _dimensions
 	width ::= this _dimensions width
 	height ::= this _dimensions height
-	elements: Float[]
+	_elements: OwnedBuffer
+	elements ::= this _elements pointer as Float*
```
Comment (davidhesselbom) [inline source/math/FloatMatrix.ooc]: So... a downside to doing things this way, that I just discovered, is that the `get` function that this syntactic sugar creates is called *every time* you access `this elements[]`.

This class has been littered with `this elements[y + x * this height]` in place of `this[x, y]` for performance reasons (by inlining the operator `[]` function and thereby accessing the array directly) and this change just nullified that optimization. Now it looks bad for no good reason, and `OwnedBuffer__OwnedBuffer__getpointer__` just bumped to the 3rd most-called function in the other project. :(

File: source/math/FloatMatrix.ooc [hunk 3]
```diff
@@ -123,12 +126,13 @@ FloatMatrix : cover {
...
 	// </summary>
 	// <returns>Return current matrix tranposed.</returns>
-	transpose: func@ -> This {
+	transpose: func -> This {
 		result := This new(this dimensions swap())
 		for (y in 0 .. this height)
 			for (x in 0 .. this width)
 				result elements[y + x * this height] = this elements[x + y * this width]
-		result
+		this free(Owner Callee)
```
Comment (davidhesselbom) [inline source/math/FloatMatrix.ooc]: Wait, what happens to the original as a result of this call?

Comment (tomasmahlberg) [inline source/math/FloatMatrix.ooc]: We're calling `free~withCriteria()`, so it will be freed if the ownership is set to `Owner Callee`, i.e. if the instance was not explicitly taken with `take()` or directly instantiated via a constructor which defaults owner to `Owner Caller`.
```ooc
free: func@ ~withCriteria (criteria: Owner) -> Bool {
	this _owner == criteria && this free()
}
```

Comment (davidhesselbom) [inline source/math/FloatMatrix.ooc]: So in the event that the owner is `Callee`, and the matrix is square, we don't need to allocate memory for a new one but can transpose in-place?

Comment (tomasmahlberg) [inline source/math/FloatMatrix.ooc]: I guess we could reuse the memory space in that case, yes.

Comment (davidhesselbom) [inline source/math/FloatMatrix.ooc]: Actually... it wouldn't have to be square at all. The transposed matrix will need exactly as much memory as the original either way. And, this applies to any of the function in this class (if we can reuse memory, we should). More than half of all `malloc` calls in the project originate from the `FloatMatrix` constructor...

Comment (davidhesselbom) [inline source/math/FloatMatrix.ooc]: We can put it in a "pull request debt" issue, though.

File: source/math/FloatMatrix.ooc (hunk 2 above)
Comment (tomasmahlberg) [inline source/math/FloatMatrix.ooc]: Should we try to do anything about this? @simonmika What's your take on this?

Comment (davidhesselbom) [inline source/math/FloatMatrix.ooc]: So I tried building in release mode and now the function is called more times than the next 7 or so functions combined. It's easily fixed, though - whenever you need `this elements` more than once in a function (especially before a loop), store it in a variable
```ooc
buffer := this elements
for (i in 0 .. width)
    buffer[x * i]
```
etc.

File: source/math/FloatMatrix.ooc (hunk 3 above)
Comment (simonmika) [inline source/math/FloatMatrix.ooc]: I would say it belongs into a future improvement milestone, it does not make anything worse, at the moment.

Comment (davidhesselbom) [inline source/math/FloatMatrix.ooc]: No, that's true.

File: source/math/FloatMatrix.ooc [hunk 4]
```diff
@@ -24,25 +25,26 @@ FloatMatrix : cover {
...
-	identity: static func@ (order: Int) -> This {
+	identity: static func (order: Int) -> This {
 		result := This new(order, order)
 		for (i in 0 .. order)
 			result elements[i + result width * i] = 1.0f
-		result
+		result give()
 	}
 
 	setVertical: func (xOffset, yOffset: Int, vector: FloatPoint3D) {
```
Comment (davidhesselbom) [inline source/math/FloatMatrix.ooc]: These checks should be in `safe` blocks.

File: source/math/FloatMatrix.ooc [hunk 5]
```diff
@@ -113,9 +116,9 @@ FloatMatrix : cover {
 	// Creates a copy of the current matrix.
 	// </summary>
 	// <returns>Return a copy of the current matrix.</returns>
-	copy: func@ -> This {
-		result := This new(this dimensions)
-		memcpy(result elements data, this elements data, this dimensions area * Float size)
+	copy: func -> This {
+		result := This new(this _elements copy(), this dimensions)
+		this free(Owner Callee)
```
Comment (davidhesselbom) [inline source/math/FloatMatrix.ooc]: Okay, can someone explain this? Copy and free?

File: source/math/FloatMatrix.ooc [hunk 6]
```diff
@@ -141,30 +145,31 @@ FloatMatrix : cover {
 		result := 0.0f
 		for (i in 0 .. this height)
 			result += this[i, i]
+		this free(Owner Callee)
```
Comment (davidhesselbom) [inline source/math/FloatMatrix.ooc]: Why... free?

File: source/math/FloatMatrix.ooc [hunk 7]
```diff
@@ -141,30 +145,31 @@ FloatMatrix : cover {
...
+				this elements[i + second * this width] = buffer
 			}
 	}
-
-	toString: func@ -> String {
+	// TODO: DEPRECATED
+	swaprows: func@ (first, second: Int) {
+		c"FloatMatrix swaprows deprecated" println()
+		this swapRows(first, second)
+	}
```
Comment (davidhesselbom) [inline source/math/FloatMatrix.ooc]: Time to die, `swaprows`. This can safely be removed.

File: source/math/FloatMatrix.ooc [hunk 8]
```diff
@@ -219,44 +224,27 @@ FloatMatrix : cover {
 	// </summary>
 	// <param name="y">The right hand column y vector of the equation system.</param>
 	// <returns>Return the least square solution to the system.</returns>
-	solve: func@ (y: This) -> This {
+	solve: func (y: This) -> This {
 		result: This
 		if (this width > this height)
 			raise("Invalid dimensions in FloatMatrix solve")
 		// TODO: This can probably be cleaned up...
```
Comment (davidhesselbom) [inline source/math/FloatMatrix.ooc]: I guess this can be removed, now.

File: source/math/FloatMatrix.ooc (hunk 4 above)
Comment (simonmika) [inline source/math/FloatMatrix.ooc]: Which checks are you referring to?

File: source/math/FloatMatrix.ooc (hunk 5 above)
Comment (simonmika) [inline source/math/FloatMatrix.ooc]: It probably is safe to change it to an either or in a later PR that focuses on in-place updates when owner is set to callee.

File: source/math/FloatMatrix.ooc (hunk 4 above)
Comment (davidhesselbom) [inline source/math/FloatMatrix.ooc]: The ones right below this line. I guess it can wait, just noticed it now.

File: source/math/FloatMatrix.ooc (hunk 6 above)
Comment (simonmika) [inline source/math/FloatMatrix.ooc]: Because thats the convention. All method calls and operators have to free all arguments including this that have the owner set to callee.

File: source/math/FloatMatrix.ooc (hunk 7 above)
Comment (simonmika) [inline source/math/FloatMatrix.ooc]: Belongs in a different PR.

File: source/math/FloatMatrix.ooc (hunk 8 above)
Comment (simonmika) [inline source/math/FloatMatrix.ooc]: What I can see there are more that can be cleaned up. Let us take it later.

File: source/math/FloatMatrix.ooc (hunk 5 above)
Comment (davidhesselbom) [inline source/math/FloatMatrix.ooc]: It can be fixed later, I'm just wondering what's going on here, if there's something I'm missing. Why would you *ever* want to return a copy and free the original instead of just returning `this`?

File: source/math/FloatMatrix.ooc (hunk 7 above)
Comment (davidhesselbom) [inline source/math/FloatMatrix.ooc]: No, it was added in this PR and it shouldn't have been.

File: source/math/FloatMatrix.ooc (hunk 6 above)
Comment (davidhesselbom) [inline source/math/FloatMatrix.ooc]: Alright, I can see why that would be useful for this function.

File: source/math/FloatMatrix.ooc (hunk 8 above)
Comment (davidhesselbom) [inline source/math/FloatMatrix.ooc]: Why fix something and leave the TODO? It here will only cause confusion, indicating there is something to be fixed here when there no longer is.

Comment (simonmika) [inline source/math/FloatMatrix.ooc]: There _are_ more things to fix.

Comment (davidhesselbom) [inline source/math/FloatMatrix.ooc]: Oh, that's what you mean. Okay then.

File: source/math/FloatMatrix.ooc (hunk 7 above)
Comment (simonmika) [inline source/math/FloatMatrix.ooc]: `swapRows` is added in this PR not `swaprows` and since it is public I have to assume that somebody uses it.

File: source/math/FloatMatrix.ooc (hunk 5 above)
Comment (simonmika) [inline source/math/FloatMatrix.ooc]: Because the normal thing is not to free.

File: source/math/FloatMatrix.ooc (hunk 7 above)
Comment (davidhesselbom) [inline source/math/FloatMatrix.ooc]: By that logic, you also can't change the signature of `lupDecomposition` or `free`.

Comment (tomasmahlberg) [inline source/math/FloatMatrix.ooc]: Sorry, I  am to blame for modifying `lupDecomposition` here, but albeit being a public function the only use of it I could find was inside the class itself, in `solve`. So it _should_ be safe.

Also, what is it that need to be sorted out before this PR can be merged? It is not entirely clear from the conversation here.

Comment (davidhesselbom) [inline source/math/FloatMatrix.ooc]: I want `swaprows` removed.
